specification/dartLangSpec.tex

```diff
@@ -14107,293 +14131,569 @@ \section{Libraries and Scripts}
 
-\rationale{
-One cannot detect the problem at compile time because compilation often occurs during execution and one does not know what the URI refers to.
-However the development environment should detect the problem.
+\rationale{%
+If two different URIs are inadvertently used to access the same library $L$,
+say, because of a symbolic link in the underlying file system,
+this may help explaining why there are two incompatible copies of
+the types declared in $L$.
+Developers may also choose to omit library names entirely,
+which motivates the exception for the empty name.%
 }
 
 \LMHash{}%
-The \Index{current library} is the library currently being compiled.
-The import modifies the namespace of the current library in a manner that is determined by the imported library and by the optional elements of the import.
+The dart core library \code{dart:core}
+is implicitly imported into every dart library other than itself
+via an import clause of the form
+\code{\IMPORT{} 'dart:core';}
+unless the importing library explicitly imports \code{dart:core}.
+Any import of \code{dart:core},
+even if restricted via \SHOW{}, \HIDE{} or \AS{},
+preempts the automatic import.
+
+\rationale{%
+It would be nice if there was nothing special about \code{dart:core}.
+However, its use is pervasive,
+which leads to the decision to import it automatically.
+However, some library $L$ may wish to define entities
+with names used by \code{dart:core}
+(which it can easily do, as the names declared by a library take precedence).
+Other libraries may wish to use $L$,
+and may want to use members of $L$ that conflict with the core library
+without having to use a prefix and without encountering errors.
+The above rule makes this possible,
+essentially canceling \code{dart:core}'s special treatment
+by means of yet another special rule.%
+}
+
+
+\subsubsection{The Imported Namespace}
+\LMLabel{theImportedNamespace}
 
 \LMHash{}%
-An immediate import directive $I$ may optionally include a prefix clause of the form \code{\AS{} \id} used to prefix names imported by $I$.
-A deferred import must include a prefix clause or a compile-time error occurs.
-It is a compile-time error if a prefix used in a deferred import is used in another import clause.
+In the following, we specify the imported namespace of a library $L$,
+\NamespaceName{\metavar{import}},
+in two steps.
 
 \LMHash{}%
-An import directive $I$ may optionally include namespace combinator clauses
-used to restrict the set of names imported by $I$.
-Currently, two namespace combinators are supported: \HIDE{} and \SHOW{}.
+We need to introduce system libraries because they have special rules.
+A \Index{system library} is a library that is part of the Dart implementation.
+Any other library is a \Index{non-system library}.
+
+\commentary{%
+A system library can generally be recognized by having
+a URI that starts with `\code{dart:}'.
+An exception is `\code{dart:ui}' which is not a system library.%
```

Comment:
Back in 2019 we agreed that it wasn't a system library, and I wanted to mention it explicitly because this situation is confusing.

Would it be a non-breaking change for 'dart:ui' to become a system library? There could be situations where it currently shadows something in a system library, and after it becomes a system library that would be a conflict.

I think we should handle that topic separately.